### PR TITLE
Add SWE-Bench + integration for concrete

### PR DESCRIPTION
**Pull request by MichaelDeng03**
- Add a concrete-inference module
  - Note: Inference being a module is kind of unnecessary. Probably should stick this into core in the future
  - __main__ and utils are copied from [swebench](https://github.com/princeton-nlp/SWE-bench/tree/main/swebench/inference), with a concrete integration added
- Add a readme for using swebench
- Add a message_fits method on OpenAIClient

Quickstart (also found in readme)
```
uv run python -m concrete-inference \
  --dataset_name_or_path princeton-nlp/SWE-bench_oracle \
  --model_name_or_path concrete \
  --output_dir ./outputs

uv run python -m swebench.harness.run_evaluation \
    --dataset_name princeton-nlp/SWE-bench \
    --predictions_path ./outputs
    --max_workers 1
    --run_id quickstart
```

Immediate TODO: Lint fix. Remove output files from git
Future TODO: Move inference into core?

### CODE REVIEW

File: benchmarks/README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# Benchmarks
+
+## SWE Bench
+
+SWE Bench in this module is copied over from the official [SWE Bench Github Page](https://github.com/princeton-nlp/SWE-bench)
+
+Several components have been removed or modified to work with the concrete framework.
```
Comment (kgang) [inline benchmarks/README.md:7]: very nit: include an addendum with a short summary of the changes

File: benchmarks/concrete-inference/concrete-inference/__main__.py [hunk 2]
```diff
@@ -0,0 +1,602 @@
+#!/usr/bin/env python3
+
+"""This python script is designed to run inference on a dataset using either the OpenAI or Anthropic API, depending on the model specified.
```
Comment (kgang) [inline benchmarks/concrete-inference/concrete-inference/__main__.py:3]: Add info on provenance and what you modified. e.g.
```suggestion
"""This file was taken from swe-bench and modified to support concrete by adding `concrete_operator_inference`.
This python script is designed to run inference on a dataset using either the OpenAI or Anthropic API, depending on the model specified. 
```

File: benchmarks/inference_outputs/concrete__SWE-bench_oracle__test.jsonl
Comment (kgang) [inline benchmarks/inference_outputs/concrete__SWE-bench_oracle__test.jsonl:1]: what is this for?

**Review by kgang — APPROVED** [review]
Looks good!
I just ran into a small bug where using the command as given with outputs being `./outputs` it failed soon after starting trying to write the .jsonl file because the folder doesn't exist
```
Generating dev split: 100%|█████████████████████████████████████████████████████| 225/225 [00:00<00:00, 6324.09 examples/s]
Generating test split: 100%|██████████████████████████████████████████████████| 2294/2294 [00:00<00:00, 9913.57 examples/s]
Generating validation split: 100%|██████████████████████████████████████████████| 191/191 [00:00<00:00, 8140.89 examples/s]
Filtered to 2294 instances
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/kentgang/git/concrete/benchmarks/concrete-inference/concrete-inference/__main__.py", line 602, in <module>
    main(**vars(args))
  File "/Users/kentgang/git/concrete/benchmarks/concrete-inference/concrete-inference/__main__.py", line 543, in main
    concrete_operator_inference(**inference_args)
  File "/Users/kentgang/git/concrete/benchmarks/concrete-inference/concrete-inference/__main__.py", line 198, in concrete_operator_inference
    with open(output_file, "a+") as f:
         ^^^^^^^^^^^^^^^^^^^^^^^
```

but it is running sloowly
```
(concrete)  π concrete michael/swebench ❯ uv run python -m concrete-inference \
  --dataset_name_or_path princeton-nlp/SWE-bench_oracle \
  --model_name_or_path concrete \
  --output_dir ./logs   
2024-12-06 14:05:40,782 INFO Will write to logs/concrete__SWE-bench_oracle__test.jsonl
2024-12-06 14:05:40,783 INFO Read 0 already completed ids from logs/concrete__SWE-bench_oracle__test.jsonl
Filtered to 2294 instances
Inference for concrete:   0%|                                                                     | 0/2294 [00:00<?, ?it/s]2024-12-06 14:05:45,973 INFO HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
Inference for concrete:   0%|                                                           | 1/2294 [00:03<2:14:03,  3.51s/it]2024-12-06 14:05:49,864 INFO HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
Inference for concrete:   0%|                                                           | 2/2294 [00:07<2:22:08,  3.72s/it]2024-12-06 14:05:53,853 INFO HTTP Request: POST https://api.openai.com/v1/chat/completions "HTTP/1.1 200 OK"
```

File: benchmarks/concrete-inference/concrete-inference/__main__.py
[the remark this answers is not in the document]
Comment (MichaelDeng03) [reply benchmarks/concrete-inference/concrete-inference/__main__.py:1]: It's a little bit of a mess and 90% copied from swebench
I figure we'll scrap most of it in the future after we figure out our own code + prompt generation.

File: benchmarks/inference_outputs/concrete__SWE-bench_oracle__test.jsonl
Comment (MichaelDeng03) [reply benchmarks/inference_outputs/concrete__SWE-bench_oracle__test.jsonl:1]: I'll delete this, and the inference_outputs folder before merging

File: benchmarks/concrete-inference/concrete-inference/__main__.py [hunk 3]
```diff
@@ -0,0 +1,602 @@
...
+    model_name_or_path (str): The name or path of the model to use.
+    output_file (str): The path to the output file.
+    model_args (dict): A dictionary of model arguments.
+    """
+    # encoding = tiktoken.encoding_for_model(model_name_or_path)
+    # test_dataset = test_dataset.filter(
+    #     lambda x: gpt_tokenize(x["text"], encoding) <= MODEL_LIMITS[model_name_or_path],
+    #     desc="Filtering",
+    #     load_from_cache_file=False,
+    # )
```
[the remark this answers is not in the document]
Comment (MichaelDeng03) [reply benchmarks/concrete-inference/concrete-inference/__main__.py:192]: Yeah, but I'm adding it to OpenAIClient.
The sorting isn't so useful, but we do need to be able to reject messages for being too long.

**Comment by kgang** [comment]
put up a real title please

File: src/concrete-core/concrete/clients/openai.py [hunk 4]
```diff
@@ -24,6 +28,46 @@ def __init__(self, model: str | None = None, temperature: float | None = None):
...
+    def message_fits(self, message: str) -> bool:
+        """
+        Check if the message fits in the model.
+        No API for checking - so this is copied manually from https://platform.openai.com/docs/models/gp
+        """
+        encoding = encoding_for_model(self.model)
+        num_tokens = len(encoding.encode(message))
+
+        # Note: No support for -latest models.
+        # Also, wtf? Why are the model names so inconsistent?
```
Comment (kgang) [inline src/concrete-core/concrete/clients/openai.py:40]: lol

**Review by kgang — APPROVED** [review]
good stuff!

**Comment by kgang** [comment]
github workflows failed from not finding concrete-inference on pip

**Pull request merged**